client/qubesdb-cmd.c

```diff
@@ -5,79 +5,135 @@
 #include <windows.h>
 #include <getopt.h>
 #else
+#include <errno.h>
 #include <unistd.h>
 #endif
 
 #include <qubesdb-client.h>
 
-int opt_fullpath = 0;
-int opt_raw = 0;
-int opt_quiet = 0;
-int opt_watch_count = 1;
-int opt_wait = 0;
+static int opt_fullpath = 0;
+static int opt_raw = 0;
+static int opt_quiet = 0;
+static int opt_watch_count = 1;
+static int opt_wait = 0;
+static int opt_rm = 0;
 
 enum {
     DO_READ = 1,
     DO_WRITE,
     DO_RM,
     DO_MULTIREAD,
     DO_LIST,
-    DO_WATCH
+    DO_WATCH,
 } qdb_cmd;
 
-void encode_and_print_value(char *val) {
+#ifdef WIN32
+#define QUBES_SECURE_ZERO SecureZeroMemory
+#else
+#define QUBES_SECURE_ZERO explicit_bzero
+#endif
+
+static int encode_and_print_value(char *val, char buffer) {
     size_t i;
-    size_t len;
+    size_t len = strlen(val);
 
     if (opt_raw) {
-        printf("%s", val);
-        return;
+        char *old_val = val;
+        i = len + 1;
+        val[len] = '\n';
+#ifndef WIN32
+        if (buffer) {
+#endif
+            if (fwrite(val, 1, i, stdout) != i) {
+                perror("fwrite");
+                QUBES_SECURE_ZERO(old_val, len + 1);
+                return 1;
+            }
+#ifndef WIN32
+        } else do {
+            ssize_t written = write(1, val, i);
+            if (written == -1) {
+                int e = errno;
+                if (e != EAGAIN && e != EWOULDBLOCK && e != EINTR) {
+                    perror("write");
+                    QUBES_SECURE_ZERO(old_val, len + 1);
+                    return 1;
+                }
+            } else {
+                val += (size_t)written;
+                i -= (size_t)written;
+            }
+        } while (i > 0);
+#endif
+        QUBES_SECURE_ZERO(old_val, len + 1);
```

Comment:
@HW42 does it make sense, for the process that is going to terminate just after print it?
This adds quite a bit of complexity (especially, _two_ ways of printing stuff, that are incompatible with each other).

Reply:
My reasoning is that I do not believe the kernel eagerly zeroes the address space of an exited process, so it is best to do so in user-space.

Reply:
I would also be fine with only using the unbuffered mode.

Reply:
@HW42 ping?

Reply:
@marmarek I agree this seems too much effort for little gain. A lot of data that ends up in RAM is sensitive, so special casing this one case doesn't seem like a good trade off. The logic consequence would also be to start mlock-ing etc.

> I would also be fine with only using the unbuffered mode.

What's the advantage?

Reply:
> @marmarek I agree this seems too much effort for little gain. A lot of data that ends up in RAM is sensitive, so special casing this one case doesn't seem like a good trade off.

CSPRNGs are expected to provide forward secrecy, in that even full system compromise does not expose values that have previously been produced and wiped from memory.  Not erasing this data breaks this property, and at least NetBSD issued a security advisory for such a bug if I recall correctly.

That said, a better approach might be to special-case the random seed on both the dom0 and qube sides.  dom0 could populate the random seed via `getrandom(2)` whenever it is requested, for example, and the qube-side daemon could write the entropy directly to `/dev/urandom` without having to pass the data via the socket.  An even nicer approach, at least for Xen, would be to use a dedicated grant table entry that dom0 fills with random bytes (`read(2)` and `getrandom(2)` should be safe against concurrent writes by the VM) and which the VM writes to `/dev/urandom` before freeing.

> The logic consequence would also be to start mlock-ing etc.

If the qube’s swap is encrypted (by being on an ephemeral volatile volume) this might not be needed.  Of course OpenBSD gets this right and encrypts all swap.

> > I would also be fine with only using the unbuffered mode.
> 
> What's the advantage?

This is a prerequisite for effective zeroization of secrets.

Reply:
[ will respond to your main argument later]

> If the qube’s swap is encrypted (by being on an ephemeral volatile volume) this might not be needed. Of course OpenBSD gets this right and encrypts all swap.

Here you are contradicting yourself. If you are worried about someone recovering a secret from RAM because it might not get overwritten until next use of that address/page, you really don't want to write a copy to disc, even if encrypted (you still have the key until shutdown).

Reply:
> [ will respond to your main argument later]

If this is not worth worrying about (and it probably is not) I will retract this.  In any case, I don’t think Linux eagerly zeroes the socket buffers used for IPC, so this isn’t worth worrying about.  I suspect `init_on_free=1` is the correct solution if one really needs this.

> > If the qube’s swap is encrypted (by being on an ephemeral volatile volume) this might not be needed. Of course OpenBSD gets this right and encrypts all swap.
> 
> Here you are contradicting yourself. If you are worried about someone recovering a secret from RAM because it might not get overwritten until next use of that address/page, you really don't want to write a copy to disc, even if encrypted (you still have the key until shutdown).

OpenBSD does not have this problem because the encryption is fine-grained and the key is securely deleted when no longer needed, at least if the Wikipedia article is correct.  Other systems do.